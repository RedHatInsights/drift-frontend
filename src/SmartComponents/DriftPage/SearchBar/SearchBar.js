import React, { Component } from 'react';
import { Form, FormGroup, TextInput } from '@patternfly/react-core';
import PropTypes from 'prop-types';
import _ from 'lodash';

export class SearchBar extends Component {
    constructor(props) {
        super(props);

        this.state = {
            filter: this.props.factFilter
        };

        this.setFactFilter = this.setFactFilter.bind(this);
    }

    componentDidUpdate(prevProps) {
        if (prevProps.factFilter !== '' && this.props.factFilter === '') {
            this.setState({ filter: this.props.factFilter });
        }
    }

    updateFactFilter = (filter) => {
        this.setState({ filter });
        this.setFactFilter(filter);
    }

    setFactFilter = _.debounce(function(filter) {
        this.props.filterByFact(filter);
    }, 250);

    checkKeyPress = (event) => {
        const { activeFactFilters, handleFactFilter } = this.props;
        const { filter } = this.state;

        if (event.key === 'Enter') {
            event.preventDefault();
            if (!activeFactFilters.includes(filter)) {
                handleFactFilter(filter);
            }
        }
    }

    render() {
        return (
            <React.Fragment>
                <Form>
                    <FormGroup
                        isRequired
                        type="text"
                        fieldId="filter"
                        onKeyPress={ this.checkKeyPress }
                    >
                        <TextInput
                            value={ this.state.filter }
                            id="filterByFact"
                            placeholder="Filter by fact"
                            onChange={ this.updateFactFilter }
                            aria-label="filter by fact"
                            data-ouia-component-type="PF4/TextInput"
                            data-ouia-component-id="fact-filter-input-comparison"
                        />
                    </FormGroup>
                </Form>
            </React.Fragment>
        );
    }
}

SearchBar.propTypes = {
    filterByFact: PropTypes.func,
    factFilter: PropTypes.string,
    handleFactFilter: PropTypes.func,
    activeFactFilters: PropTypes.array
};

export default SearchBar;
