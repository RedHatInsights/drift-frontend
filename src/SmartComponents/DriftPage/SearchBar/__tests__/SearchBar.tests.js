import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import SearchBar from '../SearchBar';

describe('SearchBar', () => {
    let props;

    beforeEach(() => {
        props = {
            factFilter: '',
            activeFactFilters: [],
            handleFactFilter: jest.fn(),
            changeFactFilter: jest.fn(),
            setHistory: jest.fn()
        };
    });

    it('should render correctly', () => {
        const wrapper = shallow(
            <SearchBar { ...props } />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should update state fact filter on componentDidUpdate', () => {
        props.factFilter = 'tripel';
        const wrapper = shallow(
            <SearchBar { ...props } />
        );

        const prevProps = wrapper.props();
        wrapper.setProps({ factFilter: '' });
        wrapper.instance().componentDidUpdate(prevProps);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should update fact filter', () => {
        props.factFilter = 'tripel';
        const wrapper = shallow(
            <SearchBar { ...props } />
        );

        wrapper.instance().updateFactFilter('dubbel');

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should catch Enter on keypress', async() => {
        props.factFilter = 'quadrupel';
        const event = { key: 'Enter', preventDefault: jest.fn() };
        const wrapper = shallow(
            <SearchBar { ...props } />
        );

        wrapper.find('FormGroup').simulate('keypress', event);

        await expect(props.handleFactFilter).toHaveBeenCalledWith('quadrupel');
        expect(props.setHistory).toHaveBeenCalled();
    });
});
