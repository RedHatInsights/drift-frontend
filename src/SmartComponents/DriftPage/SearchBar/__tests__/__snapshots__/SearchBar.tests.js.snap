// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should render correctly 1`] = `
<Fragment>
  <Form>
    <FormGroup
      fieldId="filter"
      isRequired={true}
      onKeyPress={[Function]}
      type="text"
    >
      <TextInput
        aria-label="filter by fact"
        data-ouia-component-id="fact-filter-input-comparison"
        data-ouia-component-type="PF4/TextInput"
        id="filterByFact"
        onChange={[Function]}
        placeholder="Filter by fact"
        value=""
      />
    </FormGroup>
  </Form>
</Fragment>
`;

exports[`SearchBar should update fact filter 1`] = `
<Fragment>
  <Form>
    <FormGroup
      fieldId="filter"
      isRequired={true}
      onKeyPress={[Function]}
      type="text"
    >
      <TextInput
        aria-label="filter by fact"
        data-ouia-component-id="fact-filter-input-comparison"
        data-ouia-component-type="PF4/TextInput"
        id="filterByFact"
        onChange={[Function]}
        placeholder="Filter by fact"
        value="dubbel"
      />
    </FormGroup>
  </Form>
</Fragment>
`;

exports[`SearchBar should update state fact filter on componentDidUpdate 1`] = `
<Fragment>
  <Form>
    <FormGroup
      fieldId="filter"
      isRequired={true}
      onKeyPress={[Function]}
      type="text"
    >
      <TextInput
        aria-label="filter by fact"
        data-ouia-component-id="fact-filter-input-comparison"
        data-ouia-component-type="PF4/TextInput"
        id="filterByFact"
        onChange={[Function]}
        placeholder="Filter by fact"
        value=""
      />
    </FormGroup>
  </Form>
</Fragment>
`;
